src/main/java/baritone/process/BuilderProcess.java

```diff
@@ -166,17 +166,21 @@ public boolean build(String name, File schematic, Vec3i origin) {
             return false;
         }
 
+        parsed = applyMapArtAndSelection(origin, parsed);
+
+        build(name, parsed, origin);
+        return true;
+    }
+
+    private ISchematic applyMapArtAndSelection(Vec3i origin, ISchematic schematic) {
```

Comment:
I think it's better to make the parameter an `IStaticSchematic` and then do the casting at the call sites.

Reply:
Im not sure how to do this

Reply:
Right, it's not a simple as I thought.
It is possible to do without any casting (type safety, finally)
```diff
@@ -158,7 +158,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
             return false;
         }
 
-        ISchematic parsed;
+        IStaticSchematic parsed;
         try {
             parsed = format.get().parse(new FileInputStream(schematic));
         } catch (Exception e) {
@@ -166,15 +166,18 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
             return false;
         }
 
-        parsed = applyMapArtAndSelection(origin, parsed);
+        ISchematic processed = applyMapArtAndSelection(origin, parsed);
 
-        build(name, parsed, origin);
+        build(name, processed, origin);
         return true;
     }
 
-    private ISchematic applyMapArtAndSelection(Vec3i origin, ISchematic schematic) {
+    private ISchematic applyMapArtAndSelection(Vec3i origin, IStaticSchematic parsed) {
+        ISchematic schematic;
         if (Baritone.settings().mapArtMode.value) {
-            schematic = new MapArtSchematic((IStaticSchematic) schematic);
+            schematic = new MapArtSchematic(parsed);
+        } else {
+            schematic = parsed;
         }
 
         if (Baritone.settings().buildOnlySelection.value) {
@@ -221,9 +224,9 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
                 try {
                     LitematicaSchematic schematic1 = new LitematicaSchematic(CompressedStreamTools.readCompressed(Files.newInputStream(LitematicaHelper.getSchematicFile(i).toPath())), false);
                     Vec3i correctedOrigin = LitematicaHelper.getCorrectedOrigin(schematic1, i);
-                    ISchematic schematic2 = LitematicaHelper.blackMagicFuckery(schematic1, i);
-                    schematic2 = applyMapArtAndSelection(origin, schematic2);
-                    build(name, schematic2, correctedOrigin);
+                    LitematicaSchematic schematic2 = LitematicaHelper.blackMagicFuckery(schematic1, i);
+                    ISchematic schematic3 = applyMapArtAndSelection(origin, schematic2);
+                    build(name, schematic3, correctedOrigin);
                 } catch (IOException e) {
                     logDirect("Schematic File could not be loaded.");
                 }
```
However that looks horrible so I'd also be happy with what I originally meant, which is more along the lines of the following:
```diff
@@ -166,15 +166,16 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
             return false;
         }
 
-        parsed = applyMapArtAndSelection(origin, parsed);
+        parsed = applyMapArtAndSelection(origin, (IStaticSchematic) parsed);
 
         build(name, parsed, origin);
         return true;
     }
 
-    private ISchematic applyMapArtAndSelection(Vec3i origin, ISchematic schematic) {
+    private ISchematic applyMapArtAndSelection(Vec3i origin, IStaticSchematic parsed) {
+        ISchematic schematic = parsed;
         if (Baritone.settings().mapArtMode.value) {
-            schematic = new MapArtSchematic((IStaticSchematic) schematic);
+            schematic = new MapArtSchematic(parsed);
         }
 
         if (Baritone.settings().buildOnlySelection.value) {
@@ -222,7 +223,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
                     LitematicaSchematic schematic1 = new LitematicaSchematic(CompressedStreamTools.readCompressed(Files.newInputStream(LitematicaHelper.getSchematicFile(i).toPath())), false);
                     Vec3i correctedOrigin = LitematicaHelper.getCorrectedOrigin(schematic1, i);
                     ISchematic schematic2 = LitematicaHelper.blackMagicFuckery(schematic1, i);
-                    schematic2 = applyMapArtAndSelection(origin, schematic2);
+                    schematic2 = applyMapArtAndSelection(origin, (IStaticSchematic) schematic2);
                     build(name, schematic2, correctedOrigin);
                 } catch (IOException e) {
                     logDirect("Schematic File could not be loaded.");
```

Regardless of how you do it, what I want to see is the casting happening near the creation of the schematic (i.e. same method, preferably only a few lines apart) so someone poking at this in the future will see it and not run into surprising `ClassCastException`s.

(Disclaimer regarding the diffs: I have made sure Baritone still compiles after applying them, but please don't be mad at me in case the result doesn't actually work ingame)